### PR TITLE
[formidable] Update to version 1.2

### LINKED ISSUE #51341: EventEmitterOptions is not defined correctly in @types/formidable
It seems like EventEmitterOptions is not defined correctly for @types/formidable.

The formidable docs specifically say to pass options into the constructor, not to the instance. [https://www.npmjs.com/package/formidable#formidable--incomingform](https://www.npmjs.com/package/formidable#formidable--incomingform)

The tests for @types/formidable are passing options to the instance.

[https://github.com/DefinitelyTyped/DefinitelyTyped/blob/81e930ed9983d081271ebf1978551b1e4c8b5b16/types/formidable/formidable-tests.ts#L37](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/81e930ed9983d081271ebf1978551b1e4c8b5b16/types/formidable/formidable-tests.ts#L37)

When I try to pass the options to the constructor like the formidable docs recommend, I get the following error: 
```ts
const form = new IncomingForm({ maxFileSize: 50 * 1024 * 1024, multiples: true });
```
```sh
Argument of type `{ maxFileSize: 50 * 1024 * 1024, multiples: true }` is not assignable to parameter of type 'EventEmitterOptions'. 
Object literal may only specify known properties and 'maxFileSize' does not exist in type 'EventEmitterOptions'.
```

- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I understand that `@types` package versions [do not follow SemVer](https://github.com/DefinitelyTyped/DefinitelyTyped#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library).
  - Authors: @andy-ms , @evanshortiss , @itPiligrim , @taoqf

### PR DESCRIPTION
**Pull request by martin-badin**
Closes: #51341

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by martin-badin** [comment]
@peterblazejewicz Hi. Could you please review?

**Review by peterblazejewicz — COMMENTED** [review]
I think we should take opportunity and correct this DT implementation.
This package has a long history and uses aliases,
This should be possible to use a typescript like that one:

```ts
import Formidable = require("formidable");
import { IncomingForm } from "formidable";

const formOne = new Formidable({
  captureRejections: true,
});
const formTwo = new Formidable.IncomingForm({
  captureRejections: true,
});

const formThree = new IncomingForm({
  captureRejections: true,
});

console.log(formOne.encoding);
console.log(formTwo.encoding);
console.log(formThree.encoding);
```
all of those three are corrrect and executes at runtime.

```js
var formOne = new Formidable({
    captureRejections: true,
});
var formTwo = new Formidable.IncomingForm({
    captureRejections: true,
});
var formThree = new formidable_1.IncomingForm({
    captureRejections: true,
});
```
The current DT package does not support that, as it just used direct exports.

The proper export should be something like, IMO (note, this is pseudo code):
```ts
declare class IncomingForm extends events.EventEmitter {
    ...
    parse(req: http.IncomingMessage, callback?: (err: any, fields: IncomingForm.Fields, files: IncomingForm.Files) => any): void;
}

// export alias to be declared on module
declare const alias: typeof IncomingForm;

declare namespace IncomingForm {
   
    ...

    const IncomingForm =  alias;

}

export = IncomingForm;
```

**Comment by martin-badin** [comment]
@peterblazejewicz Thank you for your review. Please could you review it again?

File: types/formidable/index.d.ts [hunk 1]
```diff
@@ -1,56 +1,447 @@
-// Type definitions for Formidable 1.0.16
-// Project: https://github.com/felixge/node-formidable/
+// Type definitions for formidable 1.1
```
Comment (peterblazejewicz) [inline types/formidable/index.d.ts:1]:
```suggestion
// Type definitions for formidable 1.2
```

File: types/formidable/index.d.ts [hunk 2]
```diff
@@ -1,56 +1,447 @@
...
 // Definitions by: Wim Looman <https://github.com/Nemo157>
+//                 Martin Badin <https://github.com/martin-badin>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
 /// <reference types="node" />
 
+import { IncomingMessage } from 'http';
+import { Stream, Transform, PassThrough } from 'stream';
+import { EventEmitter } from 'events';
+import { type } from 'os';
```
Comment (peterblazejewicz) [inline types/formidable/index.d.ts:12]: IMO, unused (and kind of risky wihtout alias in TS):
```suggestion
```
is that one used somewhere?

File: types/formidable/index.d.ts [hunk 3]
```diff
@@ -1,56 +1,447 @@
...
+import { Domain } from 'domain';
```
Comment (peterblazejewicz) [inline types/formidable/index.d.ts:13]:
```suggestion
import { Domain } from 'domain';
```
IMO, unused

File: types/formidable/tslint.json [hunk 4]
```diff
@@ -1,19 +1,6 @@
...
-        "no-var-keyword": false,
-        "object-literal-shorthand": false,
-        "only-arrow-functions": false,
-        "prefer-const": false,
-        "space-before-function-paren": false,
-        "strict-export-declare-modifiers": false,
-        "trim-file": false,
-        "triple-equals": false,
-        "whitespace": false,
         "npm-naming": [true,{"mode":"code","errors":[["NeedsExportEquals",false]]}]
```
Comment (peterblazejewicz) [inline types/formidable/tslint.json:4]: why this one is needed? (the npm package name is now the same as one published + repository is the same)

**Review by peterblazejewicz — COMMENTED** [review]
Can you keep `Files`/`Files` interfaces for backward compatiblity reasons?

File: types/formidable/index.d.ts (hunk 2 above)
Comment (martin-badin) [reply types/formidable/index.d.ts:12]: sorry, no. I will remove that

File: types/formidable/tslint.json (hunk 4 above)
Comment (martin-badin) [reply types/formidable/tslint.json:4]: there was an issue before. I will remove that.

**Comment by martin-badin** [comment]
> Can you keep `Files`/`Files` interfaces for backward compatiblity reasons?

I reverted the type

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@martin-badin thx!

**Comment by martin-badin** [comment]
Re-ping @Nemo157

**Review by peterblazejewicz — APPROVED** [review]

**Review by sandersn — APPROVED** [review]

**Comment by karlhorky** [comment]
@sandersn @peterblazejewicz @martin-badin I think this is a regression for the current version of `formidable` (`1.2.2`) for the example from the readme (see the new error with the 💥 emoji inline below).

This does not occur with `@types/formidable@1.0.32`.

```js
const http = require('http');
const formidable = require('formidable');

const server = http.createServer((req, res) => {
  if (req.url === '/api/upload' && req.method.toLowerCase() === 'post') {
    // parse a file upload
    const form = formidable({ multiples: true }); // 💥 Value of type 'typeof import("/Users/k/p/courses/node_modules/@types/formidable/index.d.ts")' is not callable. Did you mean to include 'new'?

    form.parse(req, (err, fields, files) => {
      res.writeHead(200, { 'content-type': 'application/json' });
      res.end(JSON.stringify({ fields, files }, null, 2));
    });

    return;
  }

  // show a file upload form
  res.writeHead(200, { 'content-type': 'text/html' });
  res.end(`
    <h2>With Node.js <code>"http"</code> module</h2>
    <form action="/api/upload" enctype="multipart/form-data" method="post">
      <div>Text field title: <input type="text" name="title" /></div>
      <div>File: <input type="file" name="multipleFiles" multiple="multiple" /></div>
      <input type="submit" value="Upload" />
    </form>
  `);
});

server.listen(8080, () => {
  console.log('Server listening on http://localhost:8080/ ...');
});
```

Source: https://unpkg.com/browse/formidable@1.2.2/README.md

**Comment by karlhorky** [comment]
This update also makes it impossible to import and use the internal interfaces (such as `Fields` or `Files`) from the types, as far as I understand correctly 🙁

**Comment by karlhorky** [comment]
## Workaround

Here's a patch that can be applied with [`patch-package`](https://www.npmjs.com/package/patch-package) for both of the issues above, for anyone that needs it for the latest `1.2.2` version of `formidable`:

```diff
diff --git a/node_modules/@types/formidable/index.d.ts b/node_modules/@types/formidable/index.d.ts
index 0f8b9c2..070b055 100755
--- a/node_modules/@types/formidable/index.d.ts
+++ b/node_modules/@types/formidable/index.d.ts
@@ -10,7 +10,7 @@ import { IncomingMessage } from 'http';
 import { Stream, Transform, PassThrough } from 'stream';
 import { EventEmitter } from 'events';
 
-declare namespace Formidable {
+export declare namespace Formidable {
     type BufferEncoding =
         | 'ascii'
         | 'utf8'
@@ -447,4 +447,4 @@ declare class Formidable {
     static readonly DEFAULT_OPTIONS: Formidable.Options;
 }
 
-export = Formidable;
+export = (Options) => new Formidable;
```

**Comment by peterblazejewicz** [comment]
@martin-badin please review the comments, thx!

**Comment by peterblazejewicz** [comment]
@martin-badin required change is to add merge interface with class, so both `new` and function call type are supported, so like:
```ts
interface FormidableConstructor {
    (options?: Options): Formidable;
    new (options?: Options): Formidable;
}
```
this would fill the bill, the readme sports section abut supported creational calls (but does not list constructor like call (only via alias)

**Comment by martin-badin** [comment]
@peterblazejewicz @karlhorky Hi guys. I don't see in the readme that we need the 'new' prefix. None of the examples mention this. Are you sure it is needed?

**Comment by karlhorky** [comment]
In the transitional `1.2.2` version, the Formidable team is getting users ready for the upcoming version 2 that has a new API.

They have some backwards compatibility for `1.2.1` in case you still want to use the constructor version with the `new` keyword, although not on the default export.

From the readme:

```js
const formidable = require('formidable');
const form = formidable(options);

// or
const { formidable } = require('formidable');
const form = formidable(options);

// or
const { IncomingForm } = require('formidable');
const form = new IncomingForm(options);

// or
const { Formidable } = require('formidable');
const form = new Formidable(options);
```

Source: https://github.com/node-formidable/formidable/blob/8f94fba9c683e8325da39319ae725eea7174ef7f/README.md#formidable--incomingform

**Comment by martin-badin** [comment]
@karlhorky Please, could you create PR?

**Comment by karlhorky** [comment]
Not in the next few weeks probably.

I would suggest just going with the suggestion from @peterblazejewicz and writing a test for the things in the readme.

**Comment by karlhorky** [comment]
[v2 will be released soon](https://github.com/node-formidable/formidable/issues/718#issuecomment-824201593), would be good to get the types fixed before then.

cc @sandersn maybe even an idea to revert this pull request, since the previous version of the types worked with `1.2.2` before this addition from @martin-badin and @peterblazejewicz (higher likelihood that it will continue working with v2).

**Comment by peterblazejewicz** [comment]
x-ref: #51364

**Pull request merged**